### PR TITLE
Add Mutation Testing

### PR DESCRIPTION
**Pull request by settermjd**
## Synopsis

After the recent discussion on measuring test quality, adding support for [mutation testing](https://en.wikipedia.org/wiki/Mutation_testing) seems a good choice to make. My intent here is to help us validate the tests we're writing in a meaningful, yet simple, way.

If you're not familiar with mutation testing, here's a summary of what it is:

> Mutation testing (or Mutation analysis or Program mutation) is used to design new software tests and evaluate the quality of existing software tests. Mutation testing involves modifying a program in small ways.[1] Each mutated version is called a mutant and tests detect and reject mutants by causing the behavior of the original version to differ from the mutant. This is called killing the mutant. Test suites are measured by the percentage of mutants that they kill. New tests can be designed to kill additional mutants. Mutants are based on well-defined mutation operators that either mimic typical programming errors (such as using the wrong operator or variable name) or force the creation of valuable tests (such as dividing each expression by zero). The purpose is to help the tester develop effective tests or locate weaknesses in the test data used for the program or in sections of the code that are seldom or never accessed during execution.

## Changes

This PR: 

- [x] Adds support for the [Humbug Mutation Testing](https://github.com/padraic/humbug) suite
- [x] Updates the Makefile to include a target to run mutation tests

:information_desk_person: PHPUnit was upgraded to the latest version as Humbug doesn't work well with the previous version.

### CODE REVIEW

**Comment by settermjd** [comment]
The newest version of PHPUnit requires PHP 5.6 or above. I can see if an earlier version supports PHP 5.5. But ideally it'd be great to move up to PHP 5.6 as the base version.

**Comment by localheinz** [comment]
@settermjd 

:+1: For dropping support for PHP 5.5!

**Comment by settermjd** [comment]
Well, if that's good to go, then I'll add another PR which does that.

File: composer.json [hunk 1]
```diff
@@ -26,14 +26,15 @@
     "league/pipeline": "0.1.0",
     "league/route": "dev-develop",
     "refinery29/api-output": "0.3.1",
-    "zendframework/zend-diactoros": "^1.3.3"
+    "zendframework/zend-diactoros": "^1.3.3",
+    "humbug/humbug": "^1.0.0-alpha1"
```
Comment (localheinz) [inline composer.json]: @settermjd 

Can we keep the dependencies sorted alphabetically?

```
$ composer require --dev --sort-packages humbug/humbug:^1.0.0-alpha1
```

:information_desk_person: Keeping items in a list sorted alphabetically makes it easier to find an item (or to confirm that isn't contained in the  list).

File: composer.json [hunk 2]
```diff
@@ -26,14 +26,15 @@
...
   },
   "require-dev": {
     "codeclimate/php-test-reporter": "0.2.0",
     "fzaninotto/faker": "^1.5.0",
     "phpspec/nyan-formatters": "~1.0",
     "phpspec/phpspec": "^2.4.1",
-    "phpunit/phpunit": "^4.8.23",
+    "phpunit/phpunit": "^5.0",
```
Comment (localheinz) [inline composer.json]: @settermjd 

Can we explicitly require the latest version, then?

```
$ composer require --dev --sort-packages phpunit/phpunit:^5.2.5
```

This is technically blocked by #178.

File: humbug.json.dist [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+            "src"
+        ],
+	"excludes": [
+	    "vendor"
+	]
+    },
+    "chdir": "test\/Unit",
+    "timeout": 10,
+    "logs": {
+        "text": "humbuglog.txt"
```
Comment (localheinz) [inline humbug.json.dist]: @settermjd 

Should we add this file to `.gitignore`?

File: humbug.json.dist [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+    "source": {
+        "directories": [
+            "src"
+        ],
+	"excludes": [
+	    "vendor"
+	]
```
Comment (localheinz) [inline humbug.json.dist]: @settermjd 

Looks like the indentation is garbled, doesn't it?

Also please note that we have a configuration file checked in that suggests to indent JSON files with 2 spaces, please have a look at [`.editorconfig`](https://github.com/refinery29/piston/blob/master/.editorconfig#L10-L11).

There is, by the way, a PHPStorm plugin, see https://plugins.jetbrains.com/plugin/7294.

File: Makefile [hunk 5]
```diff
@@ -16,3 +16,6 @@ test: spec unit
 
 unit: composer
 	bin/phpunit --configuration=test/Unit/phpunit.xml
+
+validate-tests:
```
Comment (localheinz) [inline Makefile]: @settermjd 

How about renaming this to `humbug`?

We could also make the `test` target depend on it:

```Makefile
test: spec unit humbug
```

File: .gitignore [hunk 6]
```diff
@@ -3,3 +3,6 @@
 /vendor/
 .idea/
 .php_cs.cache
+humblog.txt
+coverage
```
Comment (localheinz) [inline .gitignore]: @settermjd 

Do we need to exclude `coverage`?

I just ran all the tests, but the directory was never created, it seems. If we don't need it, can we remove this rule?

**Comment by localheinz** [comment]
@settermjd 

Wondering, will `humbug` ever fail a run (exit code other than `0`)?

File: composer.json (hunk 1 above)
Comment (settermjd) [inline composer.json]: sure can. I'll update in a sec.

File: humbug.json.dist (hunk 3 above)
Comment (settermjd) [inline humbug.json.dist]: erm, ok.

File: humbug.json.dist (hunk 4 above)
Comment (settermjd) [inline humbug.json.dist]: First I heard of it. Excellent stuff!

File: Makefile (hunk 5 above)
Comment (settermjd) [inline Makefile]: Nice. I didn't add it to test as it's rather new. Plus I didn't know if I was the only one familiar, or becoming familiar, with it.

**Comment by settermjd** [comment]
@localheinz honestly I don't know. I'm still new to the package, so can't say for sure. But I'll try and find out.

File: Makefile (hunk 5 above)
Comment (localheinz) [inline Makefile]: @settermjd 

:+1:

File: composer.json [hunk 7]
```diff
@@ -18,6 +18,10 @@
     {
       "name": "Ryan Catlin",
       "email": "ryan.catlin@refinery29.com"
+    },
+    {
+      "name": "Matthew Setter",
+      "email": "matthew.setter@refinery29.com"
```
Comment (localheinz) [inline composer.json]: @settermjd 

:+1:

**Comment by localheinz** [comment]
@settermjd 

:+1: Needs a rebase, then we're good to go!

**Pull request closed without merging**